### PR TITLE
Use original process spec for execs

### LINKED ISSUE #38865: Exec process doesn't preserve init process uid/gid/additionalgids since 18.06
**Description**

**Steps to reproduce the issue:**
```console
$ docker run --group-add 1234 busybox id -G
0 10 1234
$ docker run --group-add 1234 -d busybox sleep 1000
ded77fec7e0bc07542037ad4a669b7ffed3211d2762fc1d3381fa84af67f28c1
$ docker exec de id -G
0
```

**Describe the results you received:**
Exec doesn't preserve uid/gid/additionalgids.

**Describe the results you expected:**
Exec should preserve it. With 17.03.2-ce:
```
$ docker run --group-add 1234 busybox id -G
0 10 1234
lantaol@e2e-test-lantaol-minion-group-cq8p ~ $ docker run --group-add 1234 -d busybox sleep 1000
87e52362a4962267f408d6646c3de4482b5f6247de5362103c309bbeadccc175
lantaol@e2e-test-lantaol-minion-group-cq8p ~ $ docker exec 87 id -G
0 10 1234
```

**Additional information you deem important (e.g. issue happens only occasionally):**
This is introduced by https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-3ed197c0a6f7a9ca73a3ee424e448ac1.

Before updating to containerd 1.0, when init process spec is preserved by default (uid/gid/additional gids are also preserved): https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-15c000aa4859d8b3bae42251a9a13b15L61

However, after updating to containerd 1.0, the OCI `spec` is generated and passed to containerd directly https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-10e0e646b98a7523848294e49bce04ecR268, and the user field of OCI `spec` doesn't default to init process spec. (See https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-3ed197c0a6f7a9ca73a3ee424e448ac1R218 and https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-9e56cba9cecf8fa358ccc62d97eb6320R12).

**Output of `docker version`:**

```
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        4c52b90
 Built:             Wed Jan  9 19:35:23 2019
 OS/Arch:           linux/amd64
 Experimental:      true
Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 6
 Running: 3
 Paused: 0
 Stopped: 3
Images: 201
Server Version: 18.09.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 217
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 075e1ed4e0bbe6219c8ebbe57deace356970f1c7.m
runc version: 2b18fe1d885ee5083ef9f0838fee39b62d653e30
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-1021-gcp
Operating System: Ubuntu 16.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 118GiB
Name: workspace
ID: ZZQS:5I4Q:LHOK:XCBD:DOTJ:3ZTN:Y7PT:JIZI:TR6M:OQSO:INDV:WOEW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by crosbymichael**
Fixes #38865

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

### CODE REVIEW

**Review by mlaventure — APPROVED** [review]

**Comment by crosbymichael** [comment]
I have a dream that one day I'll make a PR to this repo and the tests will just pass instead of fail for some stupid reason

**Comment by cpuguy83** [comment]
Definitely seems to be some regressions after the last swarmkit vendor update.

**Comment by thaJeztah** [comment]
this is an odd one; https://jenkins.dockerproject.org/job/Docker-PRs-experimental/44458/console does that use `exec` (and no permissions because of this change?)

```
20:44:29 --- FAIL: TestTemplatedSecret (3.06s)
20:44:29     secret_test.go:221: Creating a new daemon
20:44:29     daemon.go:288: [da0323621fe41] waiting for daemon to start
20:44:29     daemon.go:320: [da0323621fe41] daemon started
20:44:29     secret_test.go:391: assertion failed: string "error injecting secret: open /go/src/github.com/docker/docker/bundles/test-integration/da0323621fe41/root/containers/e06dc8932fc969f6a5af2faa0ec942560f1d871ac9d32973318b019efaa569c5/mounts/secrets/wennbwjreg8g23oji16kxoiyj: read-only file system\r\n" does not contain "SERVICE_NAME=svc_TestTemplatedSecret\nthis is a secret\nthis is a config\n"
20:44:29     daemon.go:275: [da0323621fe41] exiting daemon
```

File: integration/container/exec_test.go [hunk 1]
```diff
@@ -117,3 +117,35 @@ func TestExec(t *testing.T) {
...
+func TestExecUser(t *testing.T) {
+	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.35"), "broken in earlier versions")
+	skip.If(t, testEnv.OSType == "windows", "FIXME. Probably needs to wait for container to be in running state.")
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true), container.WithUser("1:1"))
+
+	id, err := client.ContainerExecCreate(ctx, cID,
```
Comment (cpuguy83) [inline integration/container/exec_test.go]: Can you use `container.Exec` here?

File: daemon/exec.go [hunk 2]
```diff
@@ -217,12 +216,17 @@ func (d *Daemon) ContainerExecStart(ctx context.Context, name string, stdin io.R
...
 	}
 
-	p := &specs.Process{
-		Args:     append([]string{ec.Entrypoint}, ec.Args...),
-		Env:      ec.Env,
-		Terminal: ec.Tty,
-		Cwd:      ec.WorkingDir,
+	spec, err := d.createSpec(c)
+	if err != nil {
+		return errdefs.System(err)
```
Comment (cpuguy83) [inline daemon/exec.go]: I know the other place we call `createSpec` is wrapping this error, unfortunately it's wrong and is being addressed in another PR. Can we not wrap here?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
Error comes from here

https://github.com/moby/moby/blob/cbb885b07af59225eef12a8159e70d1485616d57/daemon/container_operations_unix.go#L209-L211

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

File: integration/container/exec_test.go [hunk 3]
```diff
@@ -117,3 +117,18 @@ func TestExec(t *testing.T) {
 	assert.Assert(t, is.Contains(out, "PWD=/tmp"), "exec command not running in expected /tmp working directory")
 	assert.Assert(t, is.Contains(out, "FOO=BAR"), "exec command not running with expected environment variable FOO")
 }
+
+func TestExecUser(t *testing.T) {
+	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.35"), "broken in earlier versions")
```
Comment (thaJeztah) [inline integration/container/exec_test.go]: oh, actually; we should adjust this version; I can backport this to 18.09 (so then it'll be `1.39`)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: integration/container/exec_test.go (hunk 3 above)
Comment (crosbymichael) [reply integration/container/exec_test.go]: What would that matter?

**Comment by cpuguy83** [comment]
:( Looks like some real CI failures.

File: integration/container/exec_test.go (hunk 3 above)
Comment (thaJeztah) [reply integration/container/exec_test.go]: (I _think_) these tests are still integrated in the internal e2e test-suite, where they run with different combinations of daemons; making sure that the version in which it works matches.

**Comment by Random-Liu** [comment]
Are we going to get this into docker 18.09.4?

**Comment by crosbymichael** [comment]
Looks like some timeouts on the CI right now

**Comment by crosbymichael** [comment]
Looks like the windows failures are legit.  looking into it today

**Comment by crosbymichael** [comment]
I think this is good to go now, rs5-process seems to be having some unrelated issues

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks 🤗

**Comment by yongtang** [comment]
All tests passed 🎉

**Pull request merged**